// fumadocs-ui/dist/components/type-table.js
'use client';

import { cn } from "../utils/cn.js";
import { Collapsible, CollapsibleContent, CollapsibleTrigger } from "./ui/collapsible.js";
import { useState } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import Link from "fumadocs-core/link";
import { ChevronDown } from "lucide-react";
import { cva } from "class-variance-authority";

//#region src/components/type-table.tsx
const keyVariants = cva("text-fd-primary", { variants: { deprecated: { true: "line-through text-fd-primary/50" } } });
const fieldVariants = cva("text-fd-muted-foreground not-prose pe-2");
function TypeTable({ type }) {
	return /* @__PURE__ */ jsxs("div", {
		className: "@container flex flex-col p-1 bg-fd-card text-fd-card-foreground rounded-2xl border my-6 text-sm overflow-hidden",
		children: [/* @__PURE__ */ jsxs("div", {
			className: "flex font-medium items-center px-3 py-1 not-prose text-fd-muted-foreground",
			children: [/* @__PURE__ */ jsx("p", {
				className: "w-[25%]",
				children: "Prop"
			}), /* @__PURE__ */ jsx("p", {
				className: "@max-xl:hidden",
				children: "Type"
			})]
		}), Object.entries(type).map(([key, value]) => /* @__PURE__ */ jsx(Item, {
			name: key,
			item: value
		}, key))]
	});
}
function Item({ name, item: { parameters = [], description, required = false, deprecated, typeDescription, default: defaultValue, type, typeDescriptionLink, returns } }) {
	const [open, setOpen] = useState(false);
	return /* @__PURE__ */ jsxs(Collapsible, {
		open,
		onOpenChange: setOpen,
		className: cn("rounded-xl border overflow-hidden transition-all", open ? "shadow-sm bg-fd-background not-last:mb-2" : "border-transparent"),
		children: [/* @__PURE__ */ jsxs(CollapsibleTrigger, {
			className: "relative flex flex-row items-center w-full group text-start px-3 py-2 not-prose hover:bg-fd-accent",
			children: [
				/* @__PURE__ */ jsxs("code", {
					className: cn(keyVariants({
						deprecated,
						className: "min-w-fit w-[25%] font-medium pe-2"
					})),
					children: [name, !required && "?"]
				}),
				typeDescriptionLink ? /* @__PURE__ */ jsx(Link, {
					href: typeDescriptionLink,
					className: "underline @max-xl:hidden",
					children: type
				}) : /* @__PURE__ */ jsx("span", {
					className: "@max-xl:hidden",
					children: type
				}),
				/* @__PURE__ */ jsx(ChevronDown, { className: "absolute end-2 size-4 text-fd-muted-foreground transition-transform group-data-[state=open]:rotate-180" })
			]
		}), /* @__PURE__ */ jsx(CollapsibleContent, { children: /* @__PURE__ */ jsxs("div", {
			className: "grid grid-cols-[1fr_3fr] gap-y-4 text-sm p-3 overflow-auto fd-scroll-container border-t",
			children: [
				/* @__PURE__ */ jsx("div", {
					className: "text-sm prose col-span-full prose-no-margin empty:hidden",
					children: description
				}),
				typeDescription && /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx("p", {
					className: cn(fieldVariants()),
					children: "Type"
				}), /* @__PURE__ */ jsx("p", {
					className: "my-auto not-prose",
					children: typeDescription
				})] }),
				defaultValue && /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx("p", {
					className: cn(fieldVariants()),
					children: "Default"
				}), /* @__PURE__ */ jsx("p", {
					className: "my-auto not-prose",
					children: defaultValue
				})] }),
				parameters.length > 0 && /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx("p", {
					className: cn(fieldVariants()),
					children: "Parameters"
				}), /* @__PURE__ */ jsx("div", {
					className: "flex flex-col gap-2",
					children: parameters.map((param) => /* @__PURE__ */ jsxs("div", {
						className: "inline-flex items-center flex-wrap gap-1",
						children: [/* @__PURE__ */ jsxs("p", {
							className: "font-medium not-prose text-nowrap",
							children: [param.name, " -"]
						}), /* @__PURE__ */ jsx("div", {
							className: "text-sm prose prose-no-margin",
							children: param.description
						})]
					}, param.name))
				})] }),
				returns && /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx("p", {
					className: cn(fieldVariants()),
					children: "Returns"
				}), /* @__PURE__ */ jsx("div", {
					className: "my-auto text-sm prose prose-no-margin",
					children: returns
				})] })
			]
		}) })]
	});
}

//#endregion
export { TypeTable };
