// fumadocs-core/dist/mdx-plugins/rehype-code.core.js
import { parseCodeBlockAttributes } from "./codeblock-utils.js";
import { t as transformerIcon } from "../transformer-icon-BYedaeE8.js";
import rehypeShikiFromHighlighter from "@shikijs/rehype/core";
import { transformerNotationDiff, transformerNotationFocus, transformerNotationHighlight, transformerNotationWordHighlight } from "@shikijs/transformers";

//#region src/mdx-plugins/rehype-code.core.ts
function rehypeCodeDefaultOptions(config) {
	return {
		lazy: true,
		...config.defaultThemes,
		defaultColor: false,
		defaultLanguage: "plaintext",
		transformers: [
			transformerNotationHighlight({ matchAlgorithm: "v3" }),
			transformerNotationWordHighlight({ matchAlgorithm: "v3" }),
			transformerNotationDiff({ matchAlgorithm: "v3" }),
			transformerNotationFocus({ matchAlgorithm: "v3" })
		],
		parseMetaString(meta) {
			const parsed = parseCodeBlockAttributes(meta, ["title", "tab"]);
			const data = parsed.attributes;
			parsed.rest = parseLineNumber(parsed.rest, data);
			data.__parsed_raw = parsed.rest;
			return data;
		}
	};
}
function parseLineNumber(str, data) {
	return str.replace(/lineNumbers=(\d+)|lineNumbers/, (_, ...args) => {
		data["data-line-numbers"] = true;
		if (args[0] !== void 0) data["data-line-numbers-start"] = Number(args[0]);
		return "";
	});
}
function createRehypeCode(configFactory) {
	return function rehypeCode(_options) {
		const { config, options } = typeof configFactory === "function" ? configFactory(_options) : {
			config: configFactory,
			options: {
				...rehypeCodeDefaultOptions(configFactory),
				..._options
			}
		};
		const transformers = options.transformers ? [...options.transformers] : [];
		transformers.unshift({
			name: "rehype-code:pre-process",
			preprocess(code, { meta }) {
				if (meta && "__parsed_raw" in meta) {
					meta.__raw = meta.__parsed_raw;
					delete meta.__parsed_raw;
				}
				if (meta && options.filterMetaString) meta.__raw = options.filterMetaString(meta.__raw ?? "");
				return code.replace(/\n$/, "");
			}
		});
		if (options.icon !== false) transformers.push(transformerIcon(options.icon));
		if (options.tab !== false) transformers.push(transformerTab());
		const transformer = Promise.resolve(config.createHighlighter()).then(async (highlighter) => {
			if ("themes" in options) await highlighter.loadTheme(...Object.values(options.themes).filter(Boolean));
			else await highlighter.loadTheme(options.theme);
			const langs = options.langs ?? (options.lazy ? [
				"js",
				"jsx",
				"ts",
				"tsx"
			] : Object.keys(highlighter.getBundledLanguages()));
			await highlighter.loadLanguage(...langs);
			return rehypeShikiFromHighlighter(highlighter, {
				...options,
				transformers
			});
		});
		return async (tree, file) => {
			await (await transformer)(tree, file, () => {});
		};
	};
}
function transformerTab() {
	return {
		name: "rehype-code:tab",
		root(root) {
			const value = this.options.meta?.tab;
			if (typeof value !== "string") return root;
			console.warn("[Fumadocs] For `tab=\"value\" in codeblocks, please use `remarkCodeTab` plugin instead.");
			return {
				type: "root",
				children: [{
					type: "mdxJsxFlowElement",
					name: "Tab",
					data: { _codeblock: true },
					attributes: [{
						type: "mdxJsxAttribute",
						name: "value",
						value
					}],
					children: root.children
				}]
			};
		}
	};
}

//#endregion
export { createRehypeCode, rehypeCodeDefaultOptions, transformerIcon, transformerTab };
