// fumadocs-ui/dist/components/sidebar/page-tree.js
import { useTreeContext, useTreePath } from "../../contexts/tree.js";
import { Fragment, useMemo } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/components/sidebar/page-tree.tsx
function createPageTreeRenderer({ SidebarFolder, SidebarFolderContent, SidebarFolderLink, SidebarFolderTrigger, SidebarSeparator, SidebarItem }) {
	function PageTreeFolder({ item, children }) {
		const path = useTreePath();
		return /* @__PURE__ */ jsxs(SidebarFolder, {
			collapsible: item.collapsible,
			active: path.includes(item),
			defaultOpen: item.defaultOpen,
			children: [item.index ? /* @__PURE__ */ jsxs(SidebarFolderLink, {
				href: item.index.url,
				external: item.index.external,
				children: [item.icon, item.name]
			}) : /* @__PURE__ */ jsxs(SidebarFolderTrigger, { children: [item.icon, item.name] }), /* @__PURE__ */ jsx(SidebarFolderContent, { children })]
		});
	}
	/**
	* Render sidebar items from page tree
	*/
	return function SidebarPageTree(components) {
		const { root } = useTreeContext();
		const { Separator, Item, Folder = PageTreeFolder } = components;
		return useMemo(() => {
			function renderSidebarList(items) {
				return items.map((item, i) => {
					if (item.type === "separator") {
						if (Separator) return /* @__PURE__ */ jsx(Separator, { item }, i);
						return /* @__PURE__ */ jsxs(SidebarSeparator, { children: [item.icon, item.name] }, i);
					}
					if (item.type === "folder") return /* @__PURE__ */ jsx(Folder, {
						item,
						children: renderSidebarList(item.children)
					}, i);
					if (Item) return /* @__PURE__ */ jsx(Item, { item }, item.url);
					return /* @__PURE__ */ jsx(SidebarItem, {
						href: item.url,
						external: item.external,
						icon: item.icon,
						children: item.name
					}, item.url);
				});
			}
			return /* @__PURE__ */ jsx(Fragment, { children: renderSidebarList(root.children) }, root.$id);
		}, [
			Folder,
			Item,
			Separator,
			root
		]);
	};
}

//#endregion
export { createPageTreeRenderer };
