// waku/dist/lib/utils/render.js
export function createRenderUtils(temporaryReferences, renderToReadableStream, createFromReadableStream, loadSsrEntryModule) {
    const onError = (e)=>{
        console.error('Error during rendering:', e);
        if (e && typeof e === 'object' && 'digest' in e && typeof e.digest === 'string') {
            return e.digest;
        }
    };
    return {
        async renderRsc (elements, options) {
            return renderToReadableStream(elements, {
                temporaryReferences,
                onError
            }, {
                onClientReference (metadata) {
                    options?.unstable_clientModuleCallback?.(metadata.deps.js);
                }
            });
        },
        async parseRsc (stream) {
            return createFromReadableStream(stream, {});
        },
        async renderHtml (elementsStream, html, options) {
            const { INTERNAL_renderHtmlStream: renderHtmlStream } = await loadSsrEntryModule();
            const rscHtmlStream = renderToReadableStream(html, {
                onError
            });
            const htmlResult = await renderHtmlStream(elementsStream, rscHtmlStream, {
                rscPath: options.rscPath,
                formState: options.formState,
                nonce: options.nonce,
                extraScriptContent: options.unstable_extraScriptContent
            });
            return new Response(htmlResult.stream, {
                status: htmlResult.status || options.status || 200,
                headers: {
                    'content-type': 'text/html'
                }
            });
        }
    };
}
