// fumadocs-ui/dist/layouts/flux/page/index.js
import { I18nLabel } from "../../../contexts/i18n.js";
import { cn } from "../../../utils/cn.js";
import { buttonVariants } from "../../../components/ui/button.js";
import { TOCProvider, TOCScrollArea } from "../../../components/toc/index.js";
import { TOCItems } from "../../../components/toc/default.js";
import { TOCItems as TOCItems$1 } from "../../../components/toc/clerk.js";
import { PageBreadcrumb, PageFooter, PageLastUpdate, PageTOCPopover, PageTOCPopoverContent, PageTOCPopoverTrigger } from "./client.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Edit } from "lucide-react";

//#region src/layouts/flux/page/index.tsx
function DocsPage({ breadcrumb: { enabled: breadcrumbEnabled = true, component: breadcrumb, ...breadcrumbProps } = {}, footer: { enabled: footerEnabled, component: footerReplace, ...footerProps } = {}, full = false, tableOfContent: { enabled: tocPopoverEnabled, component: tocPopover, ...tocOptions } = {}, toc = [], children, className }) {
	tocPopoverEnabled ??= toc.length > 0 || tocOptions.header !== void 0 || tocOptions.footer !== void 0;
	let wrapper = (children) => children;
	if (tocPopoverEnabled) wrapper = (children) => /* @__PURE__ */ jsx(TOCProvider, {
		single: tocOptions.single,
		toc,
		children
	});
	return wrapper(/* @__PURE__ */ jsxs(Fragment, { children: [tocPopoverEnabled && (tocPopover ?? /* @__PURE__ */ jsxs(PageTOCPopover, { children: [/* @__PURE__ */ jsxs(PageTOCPopoverContent, { children: [
		tocOptions.header,
		/* @__PURE__ */ jsx(TOCScrollArea, { children: tocOptions.style === "clerk" ? /* @__PURE__ */ jsx(TOCItems$1, {}) : /* @__PURE__ */ jsx(TOCItems, {}) }),
		tocOptions.footer
	] }), /* @__PURE__ */ jsx(PageTOCPopoverTrigger, {})] })), /* @__PURE__ */ jsxs("article", {
		id: "nd-page",
		"data-full": full,
		className: cn("flex flex-col w-full max-w-[900px] mx-auto [grid-area:main] px-4 py-6 gap-4 md:px-6 md:pt-8 xl:px-8 xl:pt-14", full ? "max-w-[1200px]" : "xl:layout:[--fd-toc-width:268px]", className),
		children: [
			breadcrumbEnabled && (breadcrumb ?? /* @__PURE__ */ jsx(PageBreadcrumb, { ...breadcrumbProps })),
			children,
			footerEnabled !== false && (footerReplace ?? /* @__PURE__ */ jsx(PageFooter, { ...footerProps }))
		]
	})] }));
}
function EditOnGitHub(props) {
	return /* @__PURE__ */ jsx("a", {
		target: "_blank",
		rel: "noreferrer noopener",
		...props,
		className: cn(buttonVariants({
			color: "secondary",
			size: "sm",
			className: "gap-1.5 not-prose"
		}), props.className),
		children: props.children ?? /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(Edit, { className: "size-3.5" }), /* @__PURE__ */ jsx(I18nLabel, { label: "editOnGithub" })] })
	});
}
/**
* Add typography styles
*/
function DocsBody({ children, className, ...props }) {
	return /* @__PURE__ */ jsx("div", {
		...props,
		className: cn("prose flex-1", className),
		children
	});
}
function DocsDescription({ children, className, ...props }) {
	if (children === void 0) return null;
	return /* @__PURE__ */ jsx("p", {
		...props,
		className: cn("mb-8 text-lg text-fd-muted-foreground", className),
		children
	});
}
function DocsTitle({ children, className, ...props }) {
	return /* @__PURE__ */ jsx("h1", {
		...props,
		className: cn("text-[1.75em] font-semibold", className),
		children
	});
}

//#endregion
export { DocsBody, DocsDescription, DocsPage, DocsTitle, EditOnGitHub, PageBreadcrumb, PageLastUpdate };
