// fumadocs-ui/dist/layouts/shared/language-toggle.js
'use client';

import { useI18n } from "../../contexts/i18n.js";
import { cn } from "../../utils/cn.js";
import { buttonVariants } from "../../components/ui/button.js";
import { Popover, PopoverContent, PopoverTrigger } from "../../components/ui/popover.js";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/layouts/shared/language-toggle.tsx
function LanguageToggle(props) {
	const context = useI18n();
	if (!context.locales) throw new Error("Missing `<I18nProvider />`");
	return /* @__PURE__ */ jsxs(Popover, { children: [/* @__PURE__ */ jsx(PopoverTrigger, {
		"aria-label": context.text.chooseLanguage,
		...props,
		className: cn(buttonVariants({
			color: "ghost",
			className: "gap-1.5 p-1.5"
		}), props.className),
		children: props.children
	}), /* @__PURE__ */ jsxs(PopoverContent, {
		className: "flex flex-col overflow-x-hidden p-0",
		children: [/* @__PURE__ */ jsx("p", {
			className: "mb-1 p-2 text-xs font-medium text-fd-muted-foreground",
			children: context.text.chooseLanguage
		}), context.locales.map((item) => /* @__PURE__ */ jsx("button", {
			type: "button",
			className: cn("p-2 text-start text-sm", item.locale === context.locale ? "bg-fd-primary/10 font-medium text-fd-primary" : "hover:bg-fd-accent hover:text-fd-accent-foreground"),
			onClick: () => {
				context.onChange?.(item.locale);
			},
			children: item.name
		}, item.locale))]
	})] });
}
function LanguageToggleText(props) {
	const context = useI18n();
	const text = context.locales?.find((item) => item.locale === context.locale)?.name;
	return /* @__PURE__ */ jsx("span", {
		...props,
		children: text
	});
}

//#endregion
export { LanguageToggle, LanguageToggleText };
