// fumadocs-ui/dist/components/toc/default.js
'use client';

import { useI18n } from "../../contexts/i18n.js";
import { cn } from "../../utils/cn.js";
import { mergeRefs } from "../../utils/merge-refs.js";
import { TocThumb, useTOCItems } from "./index.js";
import { useRef } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import * as Primitive from "fumadocs-core/toc";

//#region src/components/toc/default.tsx
function TOCItems({ ref, className, ...props }) {
	const containerRef = useRef(null);
	const items = useTOCItems();
	const { text } = useI18n();
	if (items.length === 0) return /* @__PURE__ */ jsx("div", {
		className: "rounded-lg border bg-fd-card p-3 text-xs text-fd-muted-foreground",
		children: text.tocNoHeadings
	});
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx(TocThumb, {
		containerRef,
		className: "absolute top-(--fd-top) h-(--fd-height) w-0.5 rounded-e-sm bg-fd-primary transition-[top,height] ease-linear"
	}), /* @__PURE__ */ jsx("div", {
		ref: mergeRefs(ref, containerRef),
		className: cn("flex flex-col border-s border-fd-foreground/10", className),
		...props,
		children: items.map((item) => /* @__PURE__ */ jsx(TOCItem, { item }, item.url))
	})] });
}
function TOCItem({ item }) {
	return /* @__PURE__ */ jsx(Primitive.TOCItem, {
		href: item.url,
		className: cn("prose py-1.5 text-sm text-fd-muted-foreground transition-colors wrap-anywhere first:pt-0 last:pb-0 data-[active=true]:text-fd-primary", item.depth <= 2 && "ps-3", item.depth === 3 && "ps-6", item.depth >= 4 && "ps-8"),
		children: item.title
	});
}

//#endregion
export { TOCItems };
