// fumadocs-ui/dist/layouts/flux/index.js
'use client';

import { cn } from "../../utils/cn.js";
import { buttonVariants } from "../../components/ui/button.js";
import { TreeContextProvider } from "../../contexts/tree.js";
import { SidebarViewport, useSidebar } from "../../components/sidebar/base.js";
import { LinkItem } from "../../utils/link-item.js";
import { SearchToggle } from "../shared/search-toggle.js";
import { renderTitleNav, useLinkItems } from "../shared/index.js";
import { LanguageToggle } from "../shared/language-toggle.js";
import { ThemeToggle } from "../shared/theme-toggle.js";
import { getSidebarTabs } from "../../components/sidebar/tabs/index.js";
import { SidebarTabsDropdown } from "./tab-dropdown.js";
import { Sidebar as Sidebar$1, SidebarContent, SidebarLinkItem, SidebarPageTree } from "./sidebar.js";
import { useMemo } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import { Languages, SidebarIcon, XIcon } from "lucide-react";
import { AnimatePresence, motion } from "motion/react";
import { RemoveScroll } from "react-remove-scroll";

//#region src/layouts/flux/index.tsx
function DocsLayout({ tree, nav = {}, sidebar: { enabled: sidebarEnabled = true, tabs: sidebarTabs, defaultOpenLevel, prefetch, ...sidebarProps } = {}, searchToggle = {}, themeSwitch = {}, i18n = false, children, containerProps, renderNavigationPanel = (props) => /* @__PURE__ */ jsx(NavigationPanel, { ...props }), ...props }) {
	const tabs = useMemo(() => {
		if (Array.isArray(sidebarTabs)) return sidebarTabs;
		if (typeof sidebarTabs === "object") return getSidebarTabs(tree, sidebarTabs);
		if (sidebarTabs !== false) return getSidebarTabs(tree);
		return [];
	}, [tree, sidebarTabs]);
	const { menuItems } = useLinkItems(props);
	const iconLinks = menuItems.filter((item) => item.type === "icon");
	function sidebar() {
		const { footer, banner, component, components, ...rest } = sidebarProps;
		if (component) return component;
		return /* @__PURE__ */ jsxs(SidebarContent, {
			...rest,
			children: [
				/* @__PURE__ */ jsx("div", {
					className: "flex flex-col gap-3 p-4 pb-2 empty:hidden",
					children: banner
				}),
				/* @__PURE__ */ jsxs(SidebarViewport, { children: [menuItems.filter((v) => v.type !== "icon").map((item, i, list) => /* @__PURE__ */ jsx(SidebarLinkItem, {
					item,
					className: cn(i === list.length - 1 && "mb-4")
				}, i)), /* @__PURE__ */ jsx(SidebarPageTree, { ...components })] }),
				footer
			]
		});
	}
	return /* @__PURE__ */ jsx(TreeContextProvider, {
		tree,
		children: /* @__PURE__ */ jsxs(Sidebar$1, {
			defaultOpenLevel,
			prefetch,
			children: [/* @__PURE__ */ jsxs("div", {
				id: "nd-flux-layout",
				...containerProps,
				className: cn("flex flex-col items-center pb-24 overflow-x-clip", containerProps?.className),
				children: [sidebarEnabled && sidebar(), children]
			}), renderNavigationPanel({
				head: renderTitleNav(nav, { className: "inline-flex items-center gap-2.5 text-sm font-semibold" }),
				tabDropdown: tabs.length > 0 && /* @__PURE__ */ jsx(SidebarTabsDropdown, {
					className: "flex-1",
					options: tabs
				}),
				tool: /* @__PURE__ */ jsxs(Fragment$1, { children: [
					i18n && /* @__PURE__ */ jsx(LanguageToggle, { children: /* @__PURE__ */ jsx(Languages, { className: "size-4.5" }) }),
					searchToggle.enabled !== false && (searchToggle.components?.sm ?? /* @__PURE__ */ jsx(SearchToggle, {
						className: "rounded-lg",
						hideIfDisabled: true
					})),
					/* @__PURE__ */ jsx(NavigationSidebarTrigger, {}),
					themeSwitch.enabled !== false && (themeSwitch.component ?? /* @__PURE__ */ jsx(ThemeToggle, {
						className: "p-1 h-full ms-1 rounded-xl bg-fd-muted *:rounded-lg",
						mode: themeSwitch.mode
					}))
				] }),
				link: iconLinks.map((item, i) => /* @__PURE__ */ jsx(LinkItem, {
					item,
					className: cn(buttonVariants({
						size: "icon-sm",
						color: "ghost"
					})),
					"aria-label": item.label,
					children: item.icon
				}, i))
			})]
		})
	});
}
function NavigationSidebarTrigger() {
	const { open, setOpen } = useSidebar();
	return /* @__PURE__ */ jsx("button", {
		className: cn(buttonVariants({
			variant: "ghost",
			size: "icon-sm",
			className: "overflow-hidden"
		})),
		onClick: () => setOpen((prev) => !prev),
		children: /* @__PURE__ */ jsx(AnimatePresence, {
			mode: "wait",
			children: /* @__PURE__ */ jsx(motion.span, {
				transition: { duration: .2 },
				initial: {
					y: "100%",
					opacity: 0
				},
				animate: {
					y: 0,
					opacity: 1
				},
				exit: {
					y: "100%",
					opacity: 0
				},
				children: open ? /* @__PURE__ */ jsx(XIcon, {}) : /* @__PURE__ */ jsx(SidebarIcon, {})
			}, open ? "open" : "closed")
		})
	});
}
function NavigationPanel({ head, tabDropdown, tool, link, className, children = (v) => v, ...props }) {
	return /* @__PURE__ */ jsx(motion.div, {
		className: cn("fixed left-1/2 w-[calc(100%-var(--removed-body-scroll-bar-size,0px))] translate-x-[calc(-50%-var(--removed-body-scroll-bar-size,0px)/2)] bottom-0 z-40 bg-fd-popover text-fd-popover-foreground border-t shadow-lg sm:bottom-6 sm:rounded-2xl sm:border sm:max-w-[380px]", className),
		...props,
		children: children(/* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsxs("div", {
			className: "flex flex-row items-center ps-2.5 p-1 gap-2 min-h-11",
			children: [head, /* @__PURE__ */ jsx("div", {
				id: "flux-layout-slot",
				className: "flex-1"
			})]
		}), /* @__PURE__ */ jsxs("div", {
			className: "flex flex-row gap-1.5 overflow-x-auto overflow-y-hidden p-2 sm:p-1",
			children: [
				/* @__PURE__ */ jsx("div", {
					className: "flex flex-row items-center gap-2 min-w-0 flex-1",
					children: tabDropdown
				}),
				/* @__PURE__ */ jsx("div", {
					className: "flex flex-row items-center text-fd-muted-foreground border-x px-0.5 empty:hidden",
					children: link
				}),
				/* @__PURE__ */ jsx("div", {
					className: "flex flex-row items-center text-fd-muted-foreground empty:hidden",
					children: tool
				})
			]
		})] }))
	});
}
function NavigationPanelOverlay({ enabled = false, className, ...props }) {
	return /* @__PURE__ */ jsx(RemoveScroll, {
		enabled,
		children: /* @__PURE__ */ jsx(motion.div, {
			className: cn("fixed inset-0 z-30 pr-(--removed-body-scroll-bar-size,0) backdrop-blur-md bg-fd-background/60", !enabled && "pointer-events-none", className),
			initial: "hide",
			variants: {
				show: { opacity: 1 },
				hide: { opacity: 0 }
			},
			animate: enabled ? "show" : "hide",
			exit: "hide",
			...props
		})
	});
}

//#endregion
export { DocsLayout, NavigationPanel, NavigationPanelOverlay };
