// fumadocs-ui/dist/layouts/flux/sidebar.js
'use client';

import { cn } from "../../utils/cn.js";
import { mergeRefs } from "../../utils/merge-refs.js";
import { SidebarCollapseTrigger as SidebarCollapseTrigger$1, SidebarFolder as SidebarFolder$1, SidebarFolderContent as SidebarFolderContent$1, SidebarFolderLink as SidebarFolderLink$1, SidebarFolderTrigger as SidebarFolderTrigger$1, SidebarItem as SidebarItem$1, SidebarProvider, SidebarSeparator as SidebarSeparator$1, SidebarTrigger as SidebarTrigger$1, SidebarViewport as SidebarViewport$1, useFolder, useFolderDepth, useSidebar } from "../../components/sidebar/base.js";
import { createLinkItemRenderer } from "../../components/sidebar/link-item.js";
import { createPageTreeRenderer } from "../../components/sidebar/page-tree.js";
import { useEffect, useEffectEvent, useRef } from "react";
import { jsx } from "react/jsx-runtime";
import { cva } from "class-variance-authority";
import { motion } from "motion/react";
import { RemoveScroll } from "react-remove-scroll";

//#region src/layouts/flux/sidebar.tsx
const MotionSidebarItem = motion.create(SidebarItem$1);
const MotionSidebarFolderTrigger = motion.create(SidebarFolderTrigger$1);
const MotionSidebarFolderLink = motion.create(SidebarFolderLink$1);
const MotionSidebarFolderContent = motion.create(SidebarFolderContent$1);
const itemVariants = cva("relative flex flex-row items-center gap-2 rounded-lg p-2 text-start text-fd-muted-foreground wrap-anywhere [&_svg]:size-4 [&_svg]:shrink-0", { variants: {
	variant: {
		link: "transition-colors hover:bg-fd-accent/50 hover:text-fd-accent-foreground/80 hover:transition-none data-[active=true]:bg-fd-primary/10 data-[active=true]:text-fd-primary data-[active=true]:hover:transition-colors",
		button: "transition-colors hover:bg-fd-accent/50 hover:text-fd-accent-foreground/80 hover:transition-none"
	},
	highlight: { true: "data-[active=true]:before:content-[''] data-[active=true]:before:bg-fd-primary data-[active=true]:before:absolute data-[active=true]:before:w-px data-[active=true]:before:inset-y-2.5 data-[active=true]:before:start-2.5" }
} });
function getItemOffset(depth) {
	return `calc(${2 + 3 * depth} * var(--spacing))`;
}
function Sidebar(props) {
	return /* @__PURE__ */ jsx(SidebarProvider, { ...props });
}
function SidebarFolder(props) {
	return /* @__PURE__ */ jsx(SidebarFolder$1, { ...props });
}
function SidebarCollapseTrigger(props) {
	return /* @__PURE__ */ jsx(SidebarCollapseTrigger$1, { ...props });
}
function SidebarViewport(props) {
	return /* @__PURE__ */ jsx(SidebarViewport$1, { ...props });
}
function SidebarTrigger(props) {
	return /* @__PURE__ */ jsx(SidebarTrigger$1, { ...props });
}
function SidebarContent({ ref: refProp, className, children, ...props }) {
	const ref = useRef(null);
	const { open, setOpen } = useSidebar();
	const listener = useEffectEvent((e) => {
		if (open && e.key === "Escape") {
			setOpen(false);
			e.preventDefault();
		}
	});
	useEffect(() => {
		window.addEventListener("keydown", listener);
		return () => {
			window.removeEventListener("keydown", listener);
		};
	}, []);
	return /* @__PURE__ */ jsx(RemoveScroll, {
		enabled: open,
		children: /* @__PURE__ */ jsx(motion.div, {
			className: cn("fixed inset-0 py-10 z-30 backdrop-blur-md bg-fd-background/60", !open && "pointer-events-none"),
			initial: "hide",
			variants: {
				show: { opacity: 1 },
				hide: { opacity: 0 }
			},
			animate: open ? "show" : "hide",
			exit: "hide",
			onClick: () => {
				setOpen(false);
			},
			children: /* @__PURE__ */ jsx(motion.div, {
				className: "absolute overflow-y-auto pr-(--removed-body-scroll-bar-size,0) [scrollbar-width:none] py-16 inset-0 bottom-26 overscroll-contain mask-[linear-gradient(to_bottom,transparent,white_calc(var(--spacing)*14),white_calc(100%-var(--spacing)*14),transparent)] lg:text-sm",
				variants: {
					show: {
						y: 0,
						opacity: 1
					},
					hide: {
						y: "80%",
						opacity: 0
					}
				},
				transition: {
					duration: .3,
					ease: [
						.16,
						1,
						.3,
						1
					],
					opacity: { duration: .1 }
				},
				children: /* @__PURE__ */ jsx(motion.aside, {
					id: "nd-sidebar",
					ref: mergeRefs(ref, refProp),
					className: cn("mx-auto sm:max-w-[400px]", className),
					onClick: (e) => e.stopPropagation(),
					...props,
					children
				})
			})
		})
	});
}
function SidebarSeparator({ className, style, children, ...props }) {
	const depth = useFolderDepth();
	return /* @__PURE__ */ jsx(SidebarSeparator$1, {
		className: cn("[&_svg]:size-4 [&_svg]:shrink-0", className),
		style: {
			paddingInlineStart: getItemOffset(depth),
			...style
		},
		...props,
		children
	});
}
function SidebarItem({ className, style, children, ...props }) {
	const depth = useFolderDepth();
	return /* @__PURE__ */ jsx(MotionSidebarItem, {
		className: cn(itemVariants({
			variant: "link",
			highlight: depth >= 1
		}), className),
		style: {
			paddingInlineStart: getItemOffset(depth),
			...style
		},
		...props,
		children
	});
}
function SidebarFolderTrigger({ className, style, ...props }) {
	const { depth, collapsible } = useFolder();
	return /* @__PURE__ */ jsx(MotionSidebarFolderTrigger, {
		className: cn(itemVariants({ variant: collapsible ? "button" : null }), "w-full", className),
		style: {
			paddingInlineStart: getItemOffset(depth - 1),
			...style
		},
		...props,
		children: props.children
	});
}
function SidebarFolderLink({ className, style, ...props }) {
	const depth = useFolderDepth();
	return /* @__PURE__ */ jsx(MotionSidebarFolderLink, {
		className: cn(itemVariants({
			variant: "link",
			highlight: depth > 1
		}), "w-full", className),
		style: {
			paddingInlineStart: getItemOffset(depth - 1),
			...style
		},
		...props,
		children: props.children
	});
}
function SidebarFolderContent({ className, children, ...props }) {
	const depth = useFolderDepth();
	const { open } = useFolder();
	return /* @__PURE__ */ jsx(MotionSidebarFolderContent, {
		className: cn("relative", depth === 1 && "before:content-[''] before:absolute before:w-px before:inset-y-1 before:bg-fd-border before:start-2.5", className),
		...props,
		children: /* @__PURE__ */ jsx(motion.div, {
			initial: "hide",
			animate: open ? "show" : "hide",
			exit: "hide",
			variants: {
				show: { opacity: 1 },
				hide: { opacity: 0 }
			},
			children
		})
	});
}
const SidebarPageTree = createPageTreeRenderer({
	SidebarFolder,
	SidebarFolderContent,
	SidebarSeparator,
	SidebarFolderLink,
	SidebarFolderTrigger,
	SidebarItem
});
const SidebarLinkItem = createLinkItemRenderer({
	SidebarFolder,
	SidebarFolderContent,
	SidebarFolderLink,
	SidebarFolderTrigger,
	SidebarItem
});

//#endregion
export { Sidebar, SidebarCollapseTrigger, SidebarContent, SidebarFolder, SidebarFolderContent, SidebarFolderLink, SidebarFolderTrigger, SidebarItem, SidebarLinkItem, SidebarPageTree, SidebarSeparator, SidebarTrigger, SidebarViewport };
